packages/devextreme/js/__internal/ui/chat/messagebox.ts

```diff
@@ -129,15 +150,54 @@ class MessageBox extends DOMComponent<MessageBox, Properties> {
   _createMessageSendAction(): void {
     this._messageSendAction = this._createActionByOption(
       'onMessageSend',
-      { excludeValidators: ['disabled', 'readOnly'] },
+      { excludeValidators: ['disabled'] },
+    );
+  }
+
+  _createTypingStartAction(): void {
+    this._typingStartAction = this._createActionByOption(
+      'onTypingStart',
+      { excludeValidators: ['disabled'] },
+    );
+  }
+
+  _createTypingEndAction(): void {
+    this._typingEndAction = this._createActionByOption(
+      'onTypingEnd',
+      { excludeValidators: ['disabled'] },
     );
   }
 
+  _triggerTypingStartAction(e: InputEvent): void {
+    if (!this._typingEndTimeoutId) {
+      this._typingStartAction?.({ event: e.event });
+    }
+  }
+
+  _updateTypingEndTimeout(): void {
+    clearTimeout(this._typingEndTimeoutId);
+
+    this._typingEndTimeoutId = setTimeout(() => {
+      this._typingEndAction?.();
+
+      this._typingEndTimeoutId = undefined;
```

Comment:
Why we don't use `_clearTypingEndTimeout` method call?

Reply:
For what? We don't need to call clearTimeout here, because after the callback is called (callback from setTimeout), clearTimeout function is called by the browser automatically. And timeout is cleared by the browser. It is strange to do it manually again.

Reply:
I mean that we have a method for these operations, and it's not a mistake to clear an already cleared timeout—the browser won't do anything in that case. This makes it easy to understand what I'm doing, instead of using just a small part of the implementation of the method.

Reply:
[2364c8ca9cda7da8629c766dd592cad9bb3c5c1a]